### ISSUE TITLE
io::copy() might be trying to use invalid file descriptor

### ISSUE THREAD

**Issue by jkarneges** [rust-lang/rust#122052]
Code:

```rust
use std::fs;

fn main() {
    fs::copy("foo", "bar").unwrap();
}
```

If I ensure "foo" exists and run the above program under valgrind on Linux, I get:

```
==75959== Warning: invalid file descriptor -1 in syscall copy_file_range(fd_in)()
```

The [docs](https://doc.rust-lang.org/std/fs/fn.copy.html#platform-specific-behavior) say `copy_file_range` may be used on Linux. Maybe it is used incorrectly? The file does get copied though.

This warning appears when the program is built with stable (1.76.0) and nightly.

**Comment by tbu-** [comment]
This seems to be intentional and not a bug:

https://github.com/rust-lang/rust/blob/80399064afa4a2cd153f30d02c25f7ea0383ed65/library/std/src/sys/pal/unix/kernel_copy.rs#L579-L581

It's probing whether the syscall `copy_file_range` exists.

**Comment by Enselic** [comment]
Triage: This this behavior seemingly is intentional, let's close as "not a bug".

**Comment by tbu-** [comment]
It was also changed in #122079 to only trigger in the error case. The example program will no longer trigger the invalid file descriptor warning.
